biblio.ttl

```diff
@@ -0,0 +1,236 @@
+@prefix derw:		<https://derwen.ai/ns/v1#> .
```

Comment:
Great question @louisguitton! 

[Diátaxis](https://diataxis.fr/) describes an antipattern, where in the case of "documentation" about W3C standards, these seem to be written for the benefit of the experts debating about the standards, i.e., mostly as *reference*. Instead there should be *tutorials*, *HOW-TOs*, *explanation*, etc., made for people trying to learn to use them. Moreover the W3C *reference* material for any single system tends to get split across many documents. Ug!

The namespace and prefix conventions in RDF are *definitely* odd, and to me they show vestiges of early HTTP, DNS, etc., in the evolution of W3C standards used for semantic graphs. I have *NEVER EVER* seen a succinct definition in print, and struggled to piece together partial answers from a variety of sources. Probably DNS is the closest analogy, though some formal notions about REST endpoints get mingled in as well.

  * _namespace_: the base portion, `https://derwen.ai/ns/v1#`, is any URL, provided as a way to distinguish a unique namespace
    - Its DNS must resolve to an actual website, and should be unique within the collection of namespaces used for an application.
    - You could use `http://example.org/#` or really whatever you want as long as you're only working locally with RDF.  Using your own domain+path is a better idea – and it might even be the domain+path for one of your GitHub repository.
    - This example ends with a hash `#` so that whatever gets follows the base URL must be handled as a [*fragment*](https://en.wikipedia.org/wiki/URI_fragment) within that namespace, i.e., some entity defined _within_ a single web page
    - Alternatively, there could be a namespace such as `http://www.wikidata.org/entity/` where whatever follows the base URL is a path, and implies there's a distinct web page for each named entity

  * _prefix_: this is a convenience provided by the "Terse Triple Language" ([TTL](https://www.w3.org/TR/turtle/)) aka "Turtle", so that RDF triples are easier to read
    - Instead of having to reference an entity "Fumay" in a "fully-qualified domain name" format: `https://derwen.ai/ns/v1#Fumay` instead I can abbreviate to `derw:Fumay`
    - These must be unique within the collection of namespaces used for an application.


So when we go to build a graph, we must mix together one or more namespaces. Each namespace references a [*controlled vocabulary*](https://en.wikipedia.org/wiki/Controlled_vocabulary) which provides shared definitions such as:

- Entities – classes used to define objects
- Relations – rules in RDFS and OWL that define possible *predicates*, i.e., relations in the graph
- Literals – data, as in "constants"

This is almost 1:1 with *object oriented programming*: imagine trying to define a class hierarchy in XML, one which allows different "flavors" of objects and multiple-inheritance.  That's pretty much what's going on here.

When you use a collection of controlled vocabularies together, that forms an *ontology* for your application. Then within the defined "classes" you being to instantiate entities and describe relations among them. Often there's one or more [*taxonomies*](https://en.wikipedia.org/wiki/Taxonomy_(general)) expressed within those relations. The taxonomy may be local to an app, or may be something shared across many applications. These are the basis for systems of classification, such as "subject headings" used in libraries, parts catalogs, etc.

In the simpler cases, where you're working with a graph in an application locally, you don't have to worry about "publishing" definitions in public, so there are less constraints.  However, often this kind of work is intended for publishing definitions online...

We use `https://github.com/hufman/flask_rdf` for Derwen, to provide a dynamic REST endpoint at that URL <https://derwen.ai/ns/v1#>   In the simplest case, it merely dumps out our public definitions (think of a public "package" on PyPi) as RDF. The REST parts of that endpoint negotiate for whatever format is required by the caller.  Then the caller can parse that response and lookup the metadata for any definition.

For instance, we publish definitions for our published content, such as `derw:topic_Natural_Language` – which you can see within the <https://derwen.ai/ns/v1#> results. That entity is part of our published taxonomy for content, which is used in several ways internally and externally. One is in our semantic search: <https://derwen.ai/d/natural_language>

The **real fun** begins to happen when you can get DNS domains, REST endpoints, content services, etc., all orchestrated on behalf of the semantics described above.  The examples that I've shown were relatively "static", although you might imagine cases where our service endpoints would provide *different* responses depending on parameters, credentials, etc.  We might publish some definitions in public, but then a superset with other private content to our clientele who authenticate and authorize through credentials.

The business use case for this is that your application which uses a graph may be consuming parts of the graph from many different sources, e.g., partner firms, clientele, public repositories, government agencies, etc. There may be reasons to "chase" the links of metadata for any specific entity in a graph, and go back to the website endpoints specified in the namespaces to lookup additional details that aren't explicitly in the graph. For instance, I really like to use Wikidata (different than DBpedia) as external references. In our public graph you'll see `wd:Q872685` which is a Wikidata definition for *data governance*, and an application could chase that down to find related topics, or translate the definitions into another language, etc.

Does this help?
Frankly, this might become a first draft for something to add to the `kglab` docs :)

Reply:
Wow, thanks a lot for this 🙏 🙏 
You might think that every time I ask the same question, but every time you provide an answer which makes something new click 😂 .

Yes this helps a lot. I like the analogy with OOP a lot. It might seem funny but dumbing it down like this helps leapfrog all the legacy of RDF conventions and "grammar" for new joiners to the KnowCon community like me.

Your parallels with the business use cases really opened my mind. I can see ways of exposing the _value_ of an ontology. A search company could expose a broad ontology for free and put a fine grained ontology behind paywall, forcing you to scroll more if you don't pay.

Lastly for WikiData, I 100% agree (shameless plug https://guitton.co/posts/wikidata/#fun-sparql-queries-related-to-football )